### ISSUE TITLE
Generated URIs have different protocol from origin request URI

### ISSUE THREAD

**Issue by bwmcbride** [epimorphics/elda#207]
In the following kind of setup:

- Browser sends a request for an html format page over https to a reverse proxy
- the proxy forwards the request over http to Elda

then some of the links on the returned page have http links, not https links, e.g. links in the sort panel and the filter panel rendered by the XSLT formatter.

Would it be possible to arrange for those links to use the same protocol as the origin browser used in the original request?

The reverse proxy and other intervening infrastructure (load balancers) can generate and maintain a header (X-forwarded-Proto) that specifies the protocol of the origin request.

**Comment by skwlilac** [comment]
I had a look to see if `api:base` could be used to do this. However I don't think that it can. The problem is in part the URI spec (possibly) and in part the java implementation of URI resolution. Internally elda does indeed hand api:base as a base URI, not merely a prefix. The problem is that a URI is not relative if it has a scheme. Setting `api:base "https:";` generates a URI parse exception because there is no scheme-specific   part following the scheme component. Setting `api:base "https:/";` resolves that problem, however the base is how a full base URI and within elda the URI resolved against it is the request path (ie. the bit to the right of the tomcat context path) which has no 'authority' field to fill in :-( so the resulting resolved URI has no authority/host field. What needs to happen inside URIUtils.resolveAgainstBase is that a base URI needs be created by resolving `api:base` against the request URI with its scheme (and possibly path,query and frag) component(s) removed so as to inject the authority component from the request into the resulting base URI if none is present in the given `api:base`. That way `api:base` can supply scheme and lhs of path (and optionally host/authority too).

Basically one want to be able to give an `api:base` that carries defaults for scheme, authority and path (lh end) and where they are missing from `api:base` they are picked up from the request URI - all prior to welding the RH end of the UI path into place.

**Comment by bwmcbride** [comment]
On 13/02/18 17:10, Stuart Williams wrote:
>
> Basically one want to be able to give an |api:base| that carries 
> defaults for scheme, authority and path (lh end) and where they are 
> missing from |api:base| they are picked up from the request URI - all 
> prior to welding the RH end of the UI path into place.
>
It would not necessarily be from the request URI, but from e.g. 
X-Forwarded-Proto header.

One, possible expensive to implement option, would be treat the api:base 
value as a template with variables as in a mod rewrite configuration.  
Predefined variable might include Host, any http header, ....  There is 
also port to consider in the general case.

Brian

**Comment by ijdickinson** [comment]
This is a duplicate of #185 

This issue has bitten me again today. Can we please prioritise fixing it?

**Comment by bwmcbride** [comment]
What is the status of this issue, given 2.0.0 has been released?
